### PR TITLE
Better retries

### LINKED ISSUE #20577: Flakiness caused by lack of retries
Seems like we don't retry downloads which can load to flakyness when they fail (and they do quite often see https://github.com/dotnet/aspnetcore/pull/20561#issuecomment-609841197). 

@dougbu also thinks this is a general problem with the scripts that we have under eng/*.

### PR DESCRIPTION
**Pull request by HaoK**
Fixes https://github.com/dotnet/aspnetcore/issues/20577
- dotnet-install.ps1 retry logic copied from eng/common/tools.ps1

### CODE REVIEW

**Comment by Pilchie** [comment]
:eyes:

**Comment by HaoK** [comment]
Gist of the change is added a new download.ps1 helper that does 5 retries with exponential backoff (copied from eng/common/tools.ps1), everywhere we were using invoke-webrequest directly has been updated to use this

**Review by dougbu — CHANGES_REQUESTED** [review]
This covers the `*.ps1` files nicely.

We also need to add `--retries 5` to all `curl` commands in `*.sh` files. Are you planning to do that in a separate PR❔

File: eng/helix/content/dotnet-install.ps1 [hunk 1]
```diff
@@ -0,0 +1,879 @@
+#
```
[the remark this answers is not in the document]
Comment (HaoK) [reply eng/helix/content/dotnet-install.ps1:1]: Oops yeah, I just was running things locally to make sure it worked, didn't notice it left a copy, good catch

File: eng/scripts/Download.ps1 [hunk 2]
```diff
@@ -0,0 +1,45 @@
+ <#
```
[the remark this answers is not in the document]
Comment (HaoK) [reply eng/scripts/Download.ps1:1]: I'll file an issue for this as there are small differences in the scripts sometimes

**Comment by HaoK** [comment]
Yeah I'll do the non ps1's in a different PR

File: eng/scripts/Download.ps1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (HaoK) [reply eng/scripts/Download.ps1:1]: https://github.com/dotnet/aspnetcore/issues/23021

File: eng/helix/content/InstallDotNet.ps1 [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    [Parameter(Mandatory = $true)]
+    $installDir    
+)
+
+& $PSScriptRoot\Download.ps1 "https://dot.net/v1/dotnet-install.ps1" $PSScriptRoot\dotnet-install.ps1
+Write-Host "Download of '$uri' complete..."
+Write-Host "Installing SDK...& $PSScriptRoot\dotnet-install.ps1 -Architecture $arch -Version $sdkVersion -InstallDir $installDir"
+Invoke-Expression "& $PSScriptRoot\dotnet-install.ps1 -Architecture $arch -Version $sdkVersion -InstallDir $installDir"
+Write-Host "Installing Runtime...& $PSScriptRoot\dotnet-install.ps1 -Architecture $arch -Runtime dotnet -Version $runtimeVersion -InstallDir $installDir"
+Invoke-Expression "& $PSScriptRoot\dotnet-install.ps1 -Architecture $arch -Runtime dotnet -Version $runtimeVersion -InstallDir $installDir"
```
[the remark this answers is not in the document]
Comment (HaoK) [reply eng/helix/content/InstallDotNet.ps1:34]: Are you referring to https://github.com/dotnet/aspnetcore/blob/1ab62e80380971a06dd1315a2b9b60c260305ea6/eng/common/tools.ps1#L264 ?  That looks like its only trying again if its a runtime install and it tries with the private feeds, I ran into this weirdness too, but the common scripts aren't running installs twice on all failures, I don't think we should mimic this logic.  If dotnet-install.sh fails, that seems like a legit failure and not something we should retry

File: eng/helix/content/Download.ps1 [hunk 4]
```diff
@@ -29,15 +29,16 @@ while($true) {
     }
     catch {
       Write-Host "Failed to download '$uri'"
-      Write-Error $_.Exception.Message -ErrorAction Continue
     }
 
     if (++$retries -le $maxRetries) {
+      Write-Warning $_.Exception.Message -ErrorAction Continue
```
Comment (dougbu) [inline eng/helix/content/Download.ps1:35]: Have you tested to be sure `$_.Exception` is available outside the `catch`❔ I doubt it is. Might need to copy the message into a variable within the `catch` block.

**Review by dougbu — CHANGES_REQUESTED** [review]
One thing to test and maybe fix. Otherwise, looks great!

**Review by dougbu — APPROVED** [review]
Ooops

**Pull request merged**